src/xccl/xccl.h

```diff
@@ -0,0 +1,477 @@
+#pragma once
+
+#define CCL_ENABLE_ZE
+#define CCL_ENABLE_SYCL
+
+#include <ATen/xpu/XPUEvent.h>
+#include <c10/core/StreamGuard.h>
+#include <c10/xpu/XPUCachingAllocator.h>
+#include <oneapi/ccl.h>
+#include <oneapi/ccl.hpp>
+#include <torch/csrc/distributed/c10d/Backend.hpp>
+#include <torch/csrc/distributed/c10d/Store.hpp>
+#include <torch/csrc/distributed/c10d/logger.hpp>
+#include <algorithm>
+#include <cstdlib>
+#include <map>
+#include <optional>
+#include <sstream>
+#include <string>
+#include <variant>
+#include <vector>
+
+#if defined(CCL_MAJOR_VERSION) &&  \
+    ((CCL_MAJOR_VERSION > 2021) || \
+     (CCL_MAJOR_VERSION == 2021) && (CCL_MINOR_VERSION >= 15))
+#define XCCL_HAS_AVG 1
+#endif // oneCCL version >= 2021.15
+
+inline std::string reduceOpToString(c10d::ReduceOp op) {
+  switch (op) {
+    case c10d::ReduceOp::SUM:
+      return "SUM";
+    case c10d::ReduceOp::PRODUCT:
+      return "PRODUCT";
+    case c10d::ReduceOp::MIN:
+      return "MIN";
+    case c10d::ReduceOp::MAX:
+      return "MAX";
+    case c10d::ReduceOp::BAND:
+      return "BAND";
+    case c10d::ReduceOp::BOR:
+      return "BOR";
+    case c10d::ReduceOp::BXOR:
+      return "BXOR";
+    case c10d::ReduceOp::AVG:
+      return "AVG";
+    case c10d::ReduceOp::PREMUL_SUM:
+      return "PREMUL_SUM";
+    default:
+      return "UNKNOWN";
+  }
+}
+
+inline bool isCCLV2EnabledCached() {
+  static const bool cachedValue = []() {
+    const char* use_ccl_v2_env = std::getenv("USE_CCL_V2");
+    if (use_ccl_v2_env) {
+      std::string value(use_ccl_v2_env);
+      std::transform(
+          value.begin(), value.end(), value.begin(), [](unsigned char c) {
+            return std::tolower(c);
+          });
+      return value == "on" || value == "yes" || value == "1";
+    }
+    return false;
+  }();
+  return cachedValue;
+}
+
+static std::string versionString = []() {
```

Comment:
```suggestion
inline const std::string& getVersionString() {
static std::string versionString = []() {

...
return versionString;
```
let's init it lazily.